### DISCUSSION TITLE
Changes for Gen 4

### DISCUSSION THREAD

**Discussion by wills106** [wills106/homeassistant-solax-modbus#16] [Ideas]
@infradom I have started to go through the changes you have made for Gen 4 and serial comms.
I'll list a few observations below as not sure where else to discus.

https://github.com/infradom/homsassistant-solax-modbus/blob/1a7482c8e52dfce2b307c3d5650e249bec6f009c/solax_modbus/__init__.py#L420-L421

Requires the value multiplying by 0.1 for Gen 2 and Gen 3

https://github.com/infradom/homsassistant-solax-modbus/blob/1a7482c8e52dfce2b307c3d5650e249bec6f009c/solax_modbus/__init__.py#L423-L424

Requires the value multiplying by 10 for Gen 2 and 0.1 for Gen 3

I suggest we use something like:

```
if self.read_gen2x1 == True or self.read_gen3x1 == True or self.read_gen3x3 == True: 
    export_factory = 0.1 
else: 
    export_factory = 0 

if self.read_gen2x1 == True: 
    export_user = 10 
elif self.read_gen3x1 == True or self.read_gen3x3 == True: 
    export_user = 0.1 
else: 
    export_user = 0

export_control_factory_limit = decoder.decode_16bit_uint()
    self.data["export_control_factory_limit"] = round(export_control_factory_limit * export_factory, 1)
        
export_control_user_limit = decoder.decode_16bit_uint()
    self.data["export_control_user_limit"] = round(export_control_user_limit * export_user, 1)
```

I know this goes against what's in the Modbus docs, but a number of parts are incorrect in them.

https://github.com/infradom/homsassistant-solax-modbus/blob/1a7482c8e52dfce2b307c3d5650e249bec6f009c/solax_modbus/__init__.py#L747
Should read `if   feedin_power > 0: self.data["grid_export"] = feedin_power` and not elf.data

**Comment by infradom** [comment]
Thanks for the feedback, I am correcting this, sorry for modifying things that I could not verify (I was just relying on the documentation)
The last error was indeed a typo, dont know how this 's' character got deleted. I know I should not push before I have tested, but my git setup is a bit weird rigth now. Furthermore, I can only test for the Gen4 X3

Thanks,

Jan

**Reply by wills106** [reply]
That's why I thought I would start taking a look at it as I know from experience things aren't always how they are written in the docs.

Edit:
Just realised if we used:

```
export_factory = 0
export_user = 0
```

We would always end up with 0, it should be 1 instead.

**Comment by infradom** [comment]
yes, I corrected this and used 1 in my update
By the way, I do receive email notifications from github

**Reply by wills106** [reply]
Do you think this is at a stage now to push the changes back and go from there? I have a number of smaller changes I want to change for Gen2 Inverters but I was waiting for you to finish your changes so it doesn't become too difficult with branching and having to backtrack some of the other changes.

**Comment by infradom** [comment]
Yes, I believe the system performs well now, and it is about time to push
the changes back.
I would however like to remove an unused text_input I added, so that this
does not confuse you.
I can perform this small cleanup tomorrow.

I also noticed that since I did a HA core upgrade, I get warnings (a lot)
of this kind :

   - Detected blocking call inside the event loop. This is causing
   stability issues. Please report issue to the custom component author for
   solax_modbus doing blocking calls at
   custom_components/solax_modbus/__init__.py, line 229: return
   self._client.read_holding_registers(address, count, **kwargs)
   - Detected blocking call inside the event loop. This is causing
   stability issues. Please report issue to the custom component author for
   solax_modbus doing blocking calls at
   custom_components/solax_modbus/__init__.py, line 235: return
   self._client.read_input_registers(address, count, **kwargs)

I believe your original version might also suffer from this, so It would be
good to check his before we push back the changes.

Thanks a lot for your work, I am very happy with the solution. I will send
you a notification tomorrow when I have removed the unused experimental
part (setting charge/discharge time windows for Gen4)

With kind regards,

Jan



On Thu, Feb 17, 2022 at 7:13 PM wills106 ***@***.***> wrote:

> Do you think this is at a stage now to push the changes back and go from
> there? I have a number of smaller changes I want to change for Gen2
> Inverters but I was waiting for you to finish your changes so it doesn't
> become too difficult with branching and having to backtrack some of the
> other changes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wills106/homsassistant-solax-modbus/discussions/16#discussioncomment-2199116>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2B23TF4REKZNGTNAK4UFLU3U3CZANCNFSM5M6CZAOA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> <wills106/homsassistant-solax-modbus/repo-discussions/16/comments/2199116@
> github.com>
>


-- 
Jan Colpaert
+32 478 882435

**Reply by wills106** [reply]
I have gone back to my code for now and don't have that error.

I'll load your version back on again tomorrow or the weekend and I'll see if I start getting the errors.

If I don't get the errors maybe it's related to the Serial side of Modbus?
I'll see how I get on and let you know.

**Comment by infradom** [comment]
Did you test this with the very latest HA version?
If so, It might indeed be related to the slower serial modbus
communication. I did not have this issue before I did the HA upgrade.
I will reverify anyway to make sure I did not break this myself.

Regards,

Jan



On Thu, Feb 17, 2022 at 9:20 PM wills106 ***@***.***> wrote:

> I have gone back to my code for now and don't have that error.
>
> I'll load your version back on again tomorrow or the weekend and I'll see
> if I start getting the errors.
>
> If I don't get the errors maybe it's related to the Serial side of Modbus?
> I'll see how I get on and let you know.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wills106/homsassistant-solax-modbus/discussions/16#discussioncomment-2199935>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2B23U7YFAAFVRRGVDKIHLU3VKALANCNFSM5M6CZAOA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> <wills106/homsassistant-solax-modbus/repo-discussions/16/comments/2199935@
> github.com>
>


-- 
Jan Colpaert
+32 478 882435

**Reply by wills106** [reply]
I am on 2022.2.5 at the moment.

I'll have a read to see what changes HA has made between 2022.2.5 and 2022.2.8

Might also be worth checking out newer pymodbus versions? We are using pymodbus==2.3.0 there is now 2.5.3

**Comment by infradom** [comment]
I am on 2022.2.6
I will do some testing with the more recent pymodbus version if I find time
tomorrow.



On Thu, Feb 17, 2022 at 9:36 PM wills106 ***@***.***> wrote:

> I am on 2022.2.5 at the moment.
>
> I'll have a read to see what changes HA has made between 2022.2.5 and
> 2022.2.8
>
> Might also be worth checking out newer pymodbus versions? We are using
> pymodbus==2.3.0 there is now 2.5.3
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wills106/homsassistant-solax-modbus/discussions/16#discussioncomment-2200035>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2B23SYDBJA3YZ6QM3KKGLU3VL4XANCNFSM5M6CZAOA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> <wills106/homsassistant-solax-modbus/repo-discussions/16/comments/2200035@
> github.com>
>


-- 
Jan Colpaert
+32 478 882435

**Comment by MJP-76** [comment]
I did see there was a modbus update recently in HA


Kind Regards

Matthew Porter

________________________________

**Comment by infradom** [comment]
I adapted my fork to the pymodbus version 2.5.3, but the warnings still
come (probably only for people with serial modbus)

I believe we are ready for a merge back into the wills106 main line, but
the number of changes is rather high. I tried not to impact the older
generations, but I cannot re-test this.
You will also see that I made the python coding style a bit more compact
for some source files, I hope you can live with this, and am sorry for
inconvenience.

With kind regards,

Jan

On Thu, Feb 17, 2022 at 10:21 PM matthewjporter ***@***.***>
wrote:

> I did see there was a modbus update recently in HA
>
>
> Kind Regards
>
> Matthew Porter
>
> ________________________________
> From: wills106 ***@***.***>
> Sent: 17 February 2022 20:36
> To: wills106/homsassistant-solax-modbus ***@***.***>
> Cc: Subscribed ***@***.***>
> Subject: Re: [wills106/homsassistant-solax-modbus] Changes for Gen 4
> (Discussion #16)
>
>
> I am on 2022.2.5 at the moment.
>
> I'll have a read to see what changes HA has made between 2022.2.5 and
> 2022.2.8
>
> Might also be worth checking out newer pymodbus versions? We are using
> pymodbus==2.3.0 there is now 2.5.3
>
> —
> Reply to this email directly, view it on GitHub<
> https://github.com/wills106/homsassistant-solax-modbus/discussions/16#discussioncomment-2200035>,
> or unsubscribe<
> https://github.com/notifications/unsubscribe-auth/ADYOAWQXWXJ2UWJKCNM4SVDU3VL4XANCNFSM5M6CZAOA
> >.
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wills106/homsassistant-solax-modbus/discussions/16#discussioncomment-2200276>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2B23VKV3S5PVEH56KYHRDU3VRGFANCNFSM5M6CZAOA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> <wills106/homsassistant-solax-modbus/repo-discussions/16/comments/2200276@
> github.com>
>


-- 
Jan Colpaert
+32 478 882435

**Comment by infradom** [comment]
I did a test today: I increased the serial baudrate from 19200 (default for Gen4) to 115200 (the maximum speed), but this did not resolve the 'blocking calls warning' issue.
I continue my search ...

**Reply by wills106** [reply]
Hopefully as more people use this integration we can get more information.

My Mum as an older rebranded SolaX PV only Inverter (I think). I'll try and get in her loft and see if it has an RS485 connection, if so I'll take a spare PI round and see if I get these blocking errors or not.

**Comment by infradom** [comment]
In the mean time, I will try to call the pymodbus read_input_registers and read_holding_registers in an async way.
I will let you know if I succeed ...

**Reply by wills106** [reply]
I managed to get round my Mum's but here inverter doesn't look like it communicates over Modbus but rather normal Uart over RS232 or RS485.

I did notice them blocking errors on a Fresh install of Home Assistant OS running on a PI4
So perhaps the calls do need to be async

Might need to call this integration SolaX Modbus & UART if I am to include both, assuming I can understand how to communicate with it?

**Comment by infradom** [comment]
Modbus RTU (serial) always uses a stack based on following layers:
 - Modbus encoding (registers/coils structure
 - UART serial communication
 - RS485 or RS232 
 So I would not talk about UART but either 
- Modbus over TCP
- Modbus RTU (serial RS485 or RS232)

The Gen4 device uses Modbus RTU over RS485, (and yes this means UART)

**Reply by wills106** [reply]
I'll send you a PM on HA. My Mum's inverter isn't Modbus by the looks.

**Comment by infradom** [comment]
Did you notice the blocking warnings on a system with TCP modbus?

I started trying to make the calls async, but did not reach a solution yet
..
I had some storm damage in the garden, so I wanted to repair this first.

I am also running my HA on a PI4, the warnings only started to come
somewhere beginning februari, when I did a software update

Jan


On Sat, Feb 26, 2022 at 4:51 PM wills106 ***@***.***> wrote:

> I managed to get round my Mum's but here inverter doesn't look like it
> communicates over Modbus but rather normal Uart over RS232 or RS485.
>
> I did notice them blocking errors on a Fresh install of Home Assistant OS
> running on a PI4
> So perhaps the calls do need to be async
>
> Might need to call this integration SolaX Modbus & UART if I am to include
> both, assuming I can understand how to communicate with it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wills106/homsassistant-solax-modbus/discussions/16#discussioncomment-2256911>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2B23WPX36MX5CQISUCTHLU5DZJPANCNFSM5M6CZAOA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> <wills106/homsassistant-solax-modbus/repo-discussions/16/comments/2256911@
> github.com>
>


-- 
Jan Colpaert
+32 478 882435

**Reply by wills106** [reply]
The only time I saw the blocking calls was when I turned on Modbus Serial.

When using TCP I have never seen blocking errors (on x86_64).

If you search for issues on HA Core on Github with blocking calls there is a few for various integrations. I think it's just warning's at the moment, not sure if they are going to block integrations that don't use async in the future or not? Not really looked into it much. They must have made a change this year.

**Comment by infradom** [comment]
Indeed,  I saw some similar reports in the HA formums for other
integrations.
I did not find a good sample solution yet. My effort to use async is
unfinished as the communication blocked completely after my first try ...
I will continue to search for a solution, but be warned, this is not my top
priority the coming days.


On Sat, Feb 26, 2022 at 8:08 PM wills106 ***@***.***> wrote:

> The only time I saw the blocking calls was when I turned on Modbus Serial.
>
> When using TCP I have never seen blocking errors (on x86_64).
>
> If you search for issues on HA Core on Github with blocking calls there is
> a few for various integrations. I think it's just warning's at the moment,
> not sure if they are going to block integrations that don't use async in
> the future or not? Not really looked into it much. They must have made a
> change this year.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wills106/homsassistant-solax-modbus/discussions/16#discussioncomment-2257698>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2B23Q6TJE7MGREPITFUG3U5EQL7ANCNFSM5M6CZAOA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> <wills106/homsassistant-solax-modbus/repo-discussions/16/comments/2257698@
> github.com>
>


-- 
Jan Colpaert
+32 478 882435

**Reply by wills106** [reply]
There is no rush. It works as it is.

Hopefully more people will start to use this and we can get more contributions.
